ethexe/contracts/src/Mirror.sol

```diff
@@ -172,14 +176,104 @@ contract Mirror is IMirror {
+                            && topic1 != Reply.selector && topic1 != ValueClaimed.selector
+                    ) {
+                        uint256 size;
+                        unchecked {
+                            size = payload.length - topicsLengthInBytes;
+                        }
+
+                        uint256 memPtr = Memory.allocate(size);
+                        assembly ("memory-safe") {
+                            calldatacopy(memPtr, add(payload.offset, topicsLengthInBytes), size)
+                        }
+
+                        bytes32 topic2;
+                        bytes32 topic3;
+                        bytes32 topic4;
+                        assembly ("memory-safe") {
+                            topic2 := calldataload(add(payload.offset, 33))
+                            topic3 := calldataload(add(payload.offset, 65))
+                            topic4 := calldataload(add(payload.offset, 97))
+                        }
+
+                        if (topicsLength == 1) {
+                            assembly ("memory-safe") {
+                                log1(memPtr, size, topic1)
+                            }
+                        } else if (topicsLength == 2) {
+                            assembly ("memory-safe") {
+                                log2(memPtr, size, topic1, topic2)
+                            }
+                        } else if (topicsLength == 3) {
+                            assembly ("memory-safe") {
+                                log3(memPtr, size, topic1, topic2, topic3)
+                            }
+                        } else if (topicsLength == 4) {
+                            assembly ("memory-safe") {
+                                log4(memPtr, size, topic1, topic2, topic3, topic4)
+                            }
+                        }
+
+                        return;
+                    }
+                }
+            }
+        }
+
         emit Message(_message.id, _message.destination, _message.payload, _message.value);
     }
 
     /// @dev Non-zero value always sent since never goes to mailbox.
     function _sendReplyMessage(Gear.Message calldata _message) private {
         _transferValue(_message.destination, _message.value);
 
-        emit Reply(_message.payload, _message.value, _message.replyDetails.to, _message.replyDetails.code);
+        if (_message.destination.code.length > 0) {
+            bytes4 replyCode = _message.replyDetails.code;
+            uint8 replyCodeDiscriminant = uint8(bytes1(replyCode));
+
+            bytes memory payload;
+
+            if (replyCodeDiscriminant == 0) {
+                /* gear_core::message::ReplyCode::Success = 0 */
+                payload = _message.payload;
+            } else {
+                /* gear_core::message::ReplyCode::{Error = 1, Unsupported = 255} */
+                payload = abi.encodeWithSelector(ICallbacks.errorReply.selector, _message.id, replyCode);
```

Comment:
should we pass other message details to `errorReply()` such as `_message.payload` (e.g. for emit event with panic string)?

Reply:
maybe it's also worth renaming method to something like `onErrorReply`?